examples/speechlm2/salm_train.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+import torch
+from lightning.pytorch import Callback, Trainer
+from omegaconf import OmegaConf, open_dict
+
+from nemo.collections.speechlm2 import SALM, DataModule, SALMDataset
+from nemo.collections.speechlm2.parts.precision import HalfPrecisionForAudio
+from nemo.core.config import hydra_runner
+from nemo.utils.exp_manager import exp_manager
+from nemo.utils.trainer_utils import resolve_trainer_cfg
+
+torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
+
+
+class PROFILING(Callback):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def on_train_batch_start(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", batch, batch_idx: int
+    ) -> None:
+        if batch_idx == 0:
+            print("STARTING PROFILE")
+            torch.cuda.profiler.cudart().cudaProfilerStart()
+
+    def on_train_batch_end(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", outputs, batch, batch_idx: int
+    ) -> None:
+        if batch_idx == 10:
+            print("STOPPING PROFILE")
+            torch.cuda.profiler.cudart().cudaProfilerStop()
+
+
+@hydra_runner(config_path="conf", config_name="salm")
+def train(cfg):
+    OmegaConf.resolve(cfg)
+    torch.distributed.init_process_group(backend="nccl")
+    torch.set_float32_matmul_precision("medium")
+    precision = cfg.trainer.get("precision", "bf16-true")
+    with open_dict(cfg):
+        cfg.trainer.pop("precision", None)
+    trainer = Trainer(
+        precision=HalfPrecisionForAudio(precision),
```

Comment:
Should this be the following?

```
trainer = Trainer(
        plugins=[HalfPrecisionForAudio(precision)],
        **resolve_trainer_cfg(cfg.trainer),
        # callbacks=[PROFILING()],
    )
```

With `precision=HalfPrecisionForAudio(precision)`, `Trainer` throws an error `ValueError: Precision <nemo.collections.speechlm2.parts.precision.HalfPrecisionForAudio object at 0x15538ac1e120> is invalid. Allowed precision values: ('transformer-engine', 'transformer-engine-float16', '16-true', '16-mixed', 'bf16-true', 'bf16-mixed', '32-true', '64-true', 64, 32,
 16, '64', '32', '16', 'bf16') `

Reply:
Thanks, I refactored this to completely hide this logic in "resolve_trainer_cfg", simplifying the trainer script and improving the reusability.